### DISCUSSION TITLE
How to conditionally use either `sqlalchemy.dialects.sqlite.insert` or `sqlalchemy.dialects.postgresql.insert` based on the dialect being used?

### DISCUSSION THREAD

**Discussion by ghost** [sqlalchemy/sqlalchemy#10151] [Usage Questions] [answered]
Hello,

I'm planning to target both SQLite and PostgreSQL. The majority of the code heavily relies on "upsert". I want to know how I can instruct SQLAlchemy to automatically use the appropriate `insert` based on the dialect?

I wish to avoid the following pattern as it creates boilerplate code that frequently appears at the beginning of my database-related modules.

Let's assume that my "base.py" appears as follows
```python

from os import getenv
from sqlalchemy import create_engine

# The connection string could be either sqlite:// or postgresql+psycopg:// depending on an environment variable.
connection_string = getenv('POSTGRESQL_CONNECTION_STRING', 'sqlite:////tmp/temp.sqlite')
engine = create_engine(connection_string)
```

"most_other_modules.py"
```python
# Each time I need to utilize "upsert," the following block of code is present at the start.

from base import engine

if engine.name == 'sqlite':
    from sqlalchemy.dialects.sqlite import insert
else:
    from sqlalchemy.dialects.postgresql import insert

# Later in the code, I use the `insert` function along with its associated methods to perform upsert operations.
insert(...).values(...).on_conflict_do_update(...)

```

**Comment by AbdealiLoKo** [comment]
While I haven't used this much ... Most of the dialect-specific DML has the dialect-agnostic DML at the main module of `sqlalchemy`.

So, I believe you are looking for `from sqlalchemy import insert` or `from sqlalchemy.sql import insert` (these both are equivalent)

**Reply by ghost** [reply]
Thanks, I tired this before I posted, it did not do what I wanted.

**Comment by zzzeek** [comment, marked as answer]
hey there -

expedient way to handle this is to import both and use the one you need at runtime:


```py
# this would be a single module in the application.   other parts of the 
# application would import it:  `from myapp.util import upsert`

from .base import engine
from sqlalchemy.dialects.sqlite import insert as sl_upsert
from sqlalchemy.dialects.postgresql import insert as pg_upsert

def upsert(table):
    if engine.name == 'sqlite':
        return sl_upsert(table)
    elif engine.name == "postgresql":
       return pg_upsert(table)
    else:
        raise NotImplementedError(f"upsert not implemented for dialect {engine.name}")

```

**Reply by ghost** [reply]
Thank you, I was hoping SQLAlchemy had some magic methods for this, but "Simple is better than complex" 🙂
